File: Java/src/main/java/parrot/Parrot.java
package parrot;

abstract class Parrot {

    protected static final double BASE_SPEED = 12.0;
    protected static final double MINIMUM_SPEED = 0.0;

    public static EuropeanParrot anEuropeanParrot() {
        return new EuropeanParrot();
    }

    public static AfricanParrot anAfricanParrot(int numberOfCoconuts) {
        return new AfricanParrot(numberOfCoconuts);
    }

    public static NailedParrot aNailedParrot() {
        return new NailedParrot();
    }

    public static NorwegianBlueParrot aNorwegianBlueParrot(double voltage) {
        return new NorwegianBlueParrot(voltage);
    }

    abstract double getSpeed();

}

File: Kotlin/src/main/kotlin/parrot/Parrot.kt
package parrot

open class Parrot {

    companion object {
        const val BASE_SPEED: Double = 12.0
        const val MINIMUM_SPEED: Double = 0.0

        fun anEuropeanParrot() = EuropeanParrot()
        fun anAfricanParrot(numberOfCoconuts: Int) = AfricanParrot(numberOfCoconuts)
        fun aNailedParrot() = NailedParrot()
        fun aNorwegianBlueParrot(voltage: Double) = NorwegianBlueParrot(voltage)
    }

}
